General Cyborg overhaul (Sandwich of buffs and nerfs)

Rukofamicom opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

- [x]  Prevents stacking of progress bars when repairing cyborgs, but makes the repair work like tend wounds does. Repairs will be continuous until completed or resources depleted with only one click.
- [x]  Gives all cyborgs a basic (20u capacity) welding tool to hopefully diversify which cyborgs people choose to play. 
- [x]  Changes the self-repair module to be an active-use ability with a two minute cooldown - Will heal 50 damage (prioritizing burn) over 10 seconds at the cost of 500 cell charge.
- [x]  Removes flashes from all Nanotrasen cyborgs except Peacekeeper and Standard modules
- [x] Cyborgs will require two new "flashlight cameras" to build ~~instead of flashes.~~ *(they are actually still just flashes to retain all of the same general characteristics)*
- [x]  Causes cyborgs to progressively slow down after first module goes offline, reaching their slowest as the third goes offline
- [x]  Causes cyborgs without power cell charge to behave as if they were in their most damaged state
- [x]  Changes how EMPs affect cyborgs: They now do burn damage instead of brute, and also now stun for substantially less time (1.5 or 3 seconds depending on severity). EMPs also now affect cyborg radios to cut off their ability to communicate, do twice as much damage (20-40) and unequip all active modules.
- [x] Changes how EMPs affect radios (all radios, not just cyborgs): Radios are now disabled for two minutes instead of 20 seconds, but no longer have their encryption keys destroyed. 
- [x]  Changes how Flashes affect cyborgs: The blinding and stunning durations have been reversed - They now completely blind for 7 seconds and stun for 2 instead of blinding for 2 and stunning for 7.
- [x]  Remove laser pointer's massive 10-20 second stun. It will still trigger the same blinding effect as a flash for seven seconds. 
- [x] Most emagged modules are balanced around cyborgs being instantly shutdown by specific mechanics. Gonna have to weaken some of those too.
- [x]  Gives cyborg hyposprays a very brief use time so that it doesn't work on actively moving targets, but will still work quickly. 
- [x] Add some standard flashes to robotics

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Rock-paper-scissors combat is not a great model to go by. You either have the stuns cyborgs are weak to and abuse them, or you're fighting a very steep uphill battle. It's no fun for the cyborgs to be totally powerless against their weaknesses and it's not fun to need those weaknesses to really stand a chance. The presence or lack of specific weakness is no longer such a major factor in combat with cyborgs that it is almost the only factor which matters.

This smooths the curve on both sides - Cyborgs retain their weaknesses, but are less weak to them and cyborgs have also had some of their greatest strengths filed down to more reasonable levels. 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
-->

## Testing Photographs and Procedure
<!-- Include any screenshots/videos/debugging steps of the modified code functioning successfully, ideally including edge cases. -->

### Cyborg repairing to full from 150 damage
Sparks shown at the beginning of videos are attempts to stack the bars.

https://user-images.githubusercontent.com/9547572/167260955-de9b260c-3b02-4fc2-ae30-d579dff660d0.mp4

https://user-images.githubusercontent.com/9547572/167261057-3b37e3b8-db98-4413-8a5c-1a4a4b6ad5e5.mp4

### Self repair module with functional cooldown
I can't really showcase it working now that debug messages have been removed, but I did thoroughly test it with debug messages reporting each step as it happened to ensure healing was being prioritized as stated. 5 healing per step. 

https://user-images.githubusercontent.com/9547572/167261309-8a7e76b5-27f8-4d7c-9343-8305e97b7eea.mp4

![image](https://user-images.githubusercontent.com/9547572/167261503-7806b079-4fb8-49b5-ad6d-06ad87def9f7.png)
![image](https://user-images.githubusercontent.com/9547572/167261573-cfb45530-a23d-4f12-9110-1e9b4c958d70.png)

### Cyborg Eye + Flash examines, installations and removals from cyborg heads

Burnt out bulbs no longer say "it looks like it can do about 0 more flashes", and instead help players understand how to replace them. 
![image](https://user-images.githubusercontent.com/9547572/167261700-43919fc6-d400-4161-a110-81e30892129a.png)

If the bulb only has one use remaining it reads like this, which is flavorfully suitable for both flashes and cyborg eyes
![image](https://user-images.githubusercontent.com/9547572/167261730-269a8641-4758-4d80-9e96-db06679f88aa.png)

Cyborg heads now tell you they are expecting cyborg eyes on examine
![image](https://user-images.githubusercontent.com/9547572/167261779-c2d6a4d2-7c5a-4062-8aae-d7ac5dca87f5.png)
But they also tell you exactly what's in the socket since it still accepts both
![image](https://user-images.githubusercontent.com/9547572/167261910-863339ec-c132-4ced-ae03-e672cd02b26a.png)

Contextual flash messages so that using a cyborg eye doesn't reference it as a flash:
![image](https://user-images.githubusercontent.com/9547572/167261826-0d1d0f36-99a3-4588-a11b-3f7702892549.png)
![image](https://user-images.githubusercontent.com/9547572/167261845-cedd1a5c-2863-45ff-baa5-3db6cb07c3e7.png)

### Cyborg speed slowdowns

https://user-images.githubusercontent.com/9547572/167262063-b4376029-3e10-4537-b447-aaf28d35fa80.mp4

### Cyborg Hypospray

https://user-images.githubusercontent.com/9547572/167262146-14b6ab34-f39e-4e2d-ac38-f3393f488d39.mp4

### EMP and flash disruption. 
EMP is shown but the duration of the disruptive static hud is identical regardless of the stun duration. Showing a video of each would just be the same video three times with a slightly different length of blacked out time.

https://imgur.com/a/IgcaJWg

Additionally showing they do burn damage and the tell for when the radio starts working again
![image](https://user-images.githubusercontent.com/9547572/167262493-bea09f57-3f49-44ba-8cad-23a5d9de1943.png)
![image](https://user-images.githubusercontent.com/9547572/167262515-3062459f-322e-450a-92b7-d2693cc0fcd2.png)

## Changelog
:cl:
balance: Cyborg repair is no longer a spammable action. Cyborg repairs will repeat automatically until finished or resources are exhausted with one click. Cyborgs also require assistance to repair now and are unable to use their welding tools to repair themselves.
balance: All cyborgs (including syndicate) that didn't already have a welder now start with a 20u capacity welder for repairing themselves.
tweak: The self-repair module available to all cyborgs has been transformed into a burst healing ability which prioritizes burn damage with a two minute cooldown.
tweak: The "weak" flashes available to robotics for building cyborgs have been reskinned into Cyborg Eyes and reduced to one charge. The robotics lab now has a few standard flashes on every map in case cyborgs get unruly. 
balance: Engineering, Medical, Mining, Service and Clown Cyborgs have lost their Flash module. The module is now only available to the modules intended to take part in crew suppression: Peacekeeper and Standard. Syndicate cyborgs have retained their flashes. 
balance: Cyborgs now progressively slow down as they receive damage, starting from when the first module goes offline and capping when the last module goes offline. Cyborgs with power in their cells will move at the same speed as a cyborg with all modules offline due to damage. Both penalties stack together if applicable.
balance: EMPs and Flashes now stun cyborgs for much less time: 3 seconds for strong EMP, 2 seconds for flash, 1.5 seconds for weak EMP. Laser pointers no longer stun cyborgs at all.
balance: EMPs unequip all active modules on cyborgs now. 
balance: EMPs, Flashes and Laser pointers now blind cyborgs for 7 seconds, up from 2. This is a total blindness and does not have a small ring of vision directly around the player - they will be truly blind and unable to see in which direction an adjacent attacker is located. 
balance: EMPs now disable radios for two minutes, cyborgs included. 
balance: Cyborg hyposprays now have a very short use time so that they will not function on actively moving targets.
balance: Emagged Janitor cyborg has had their acid spray upgraded to Fluorosulfuric acid, to bring it slightly closer to other emagged modules. 
balance: Emagged clown cyborgs now have 3u superlube instead of 3u cfl3 in their hacked flower. Honk Honk.
balance: Emagged standard cyborgs will find that their energy sword now costs 150 cell charge to use. 
balance: Emagged Engineering and Mining cyborgs will find that their stun arm now obeys the cooldown that is supposed to be present on all melee weapons, and also that it costs 300 cell charge to use. Its stamina damage was also slightly reduced to match stun batons. 
balance: Emagged service borgs will find their shaker stocked with Bastion Bourbon, Zombie powder, and Smoke Powder (with an activator) instead of the now removed fake beer.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Rukofamicom committed:
No more progress bar stacking for borgo healing

Rukofamicom committed:
Update _basemap.dm

Rukofamicom committed:
Now works like Tend Wounds

Rukofamicom committed:
Removes checks for toxin damage

Cyborgs are completely incapable of receiving toxin damage, even via admin tools. There's no reason to check for it.

Rukofamicom committed:
return

Rukofamicom committed:
Fixes fuel consumption

Rukofamicom committed:
Cyborg welding tool no longer doubles in speed

Rukofamicom committed:
update borg welding tool description

zeskorion commented:
just remove them

Rukofamicom committed:
Welding tool to all cyborg modules

Including the Security and Syndicate borgs, but excluding the Borgi. Nobody would ever hurt a Borgi anyway.

Rukofamicom commented:
> just remove them

The only silicon I want to remove is roundstart AI. Not even AI altogether, just roundstart AI. I think it should be a project the station has to put together instead of having a 50% chance to be an oppressive big brother for every single antagonist on the station. That and a station without an active AI and no plan to make one has an absolutely fantastic private location to take over as an antagonist. 

But I feel like removing roundstart AI is a longshot. It's also not going to happen in this PR.

Rukofamicom committed:
Update robot_modules.dm

Rukofamicom committed:
Borg hypo 0.5s delay

AnCopper commented:
The laser pointer effect on cyborgs is basically just a long-range flash.

Rukofamicom committed:
Turns weak flashes into cyborg eyes

Also reduces them to one charge
Also also does a major overhaul on their how their examine text generates

Rukofamicom committed:
fix wacky desc shenanigans

Rukofamicom committed:
Assembling a head without strange words and naming issues

So that people who are putting actual flashes into cyborg heads aren't getting mismatched messages.

Rukofamicom committed:
Skeleton ability with functional cooldown

It does almost nothing currently.

Rukofamicom committed:
Skeleton ability with functional healing

Still full of debug messages and needs to be finished, but I need sleep before I continue.

Wilchenx commented:
Borgs are literally fine

Rukofamicom committed:
Sound effects and misc

Rukofamicom committed:
Reduce volume

Also removes partial refund calc. Always full cooldown after use.

Rukofamicom committed:
Final touches on repair module

Rukofamicom committed:
Removes debug messages and restores the original cooldown

Rukofamicom committed:
Slowdown without power cell

Rukofamicom commented:
> The laser pointer effect on cyborgs is basically just a long-range flash.

Finally got to this point and holy shit no it isn't. Laser pointers are FAR worse than flashes.  Both items call the flash proc, but:
* Flash is a flat 7 second stun
* Upgraded laser pointers are 10-20 second stun, depending on how the RNG lands

Rukofamicom committed:
Flash and lasers

Reduces stun time to 2 seconds and blinding static overlay to 7.5 seconds. Also changes the overlay to not be so transparent.

Rukofamicom committed:
First pass of emagged stuff

Still need to hit service borg's ludicrous fake beer.

ToasterBan commented:
some changes I would want to cyborgs is make it so yellow core slime cells don't work on them, infinite power cells were a bad idea in the first place

ZombiesVsPlants commented:
If flashes don't stun borgs, how would an ai stop a rogue borg from waltzing into the sat, tanking the laser turrets and then turning the Ai APC off?

Rukofamicom commented:
> Greatly reduced stun time for laser pointer/flash on a cyborg, but the blinding static overlay will remain for the original duration of a flash stun.

Greatly reduced stun is not the same as removing the stun completely - it's being reduced to 2 seconds so that there is less chance for follow up attacks without chain-flashing and burning the bulb out quickly. 2 seconds is the same length of a time a flash is on cooldown for. 

Additionally they are blinded by static (not the blind overlay which still lets you see a little) for the original duration of the stun: 7 seconds.

Rukofamicom commented:
> some changes I would want to cyborgs is make it so yellow core slime cells don't work on them, infinite power cells were a bad idea in the first place

I agree, but this will affect far more than just cyborgs. I like self-charging batteries in the form of slime cores, but they charge so fast they feel more like infinite power than just self-charging. I'll look at them, but not in this PR I don't think. Xenobio has a whole mess of issues.

ZombiesVsPlants commented:
Ah I see, but that raises even more questions.

How is anyone meant to subvert a borg? Either via law board or emag you can't feasibly do it in the 2 seconds you have.
Or what about rogue borgs? Unless you have someone else flashing for you I don't see how you could quickly disable it via a roboticist opening and cutting wires, and even if you did have a fellow teammate, you have 8 seconds with a standard flash (last time I played roboticist flashes had 4 charges) to get to the borg, swipe it with your ID, crowbar it open, remove the cell, screwdriver the borg and cut the lockdown wire. It just doesn't seem feasible in the limited time and I worry a lot more borgs will simply be beaten to death instead of manually locked down.

Have you accounted for Cult stun? It's meant to be a heavy EMP on borgs but if it just cuts off comms and flashes the screen cyborgs will be much more effective at dealing with cults as opposed to the cautious approach they have now due to the dangerous EMP spells.

As for the laser pointer, I made this PR on austation which seems to work well, though once the players found the formula and combined with the flash it was still a bit strong but much better than before:
https://github.com/austation/austation/pull/1845

Rukofamicom commented:
> How is anyone meant to subvert a borg? Either via law board or emag you can't feasibly do it in the 2 seconds you have.

Between the two of us, someone is either over or underestimating the effects of being totally blind. And I do mean totally blind, not "still able to see, but only half of directly adjacent spaces". Cyborg sensory overload is represented as an absolute, unlike a blind person who is still able to feel their immediate surroundings. 

![image](https://user-images.githubusercontent.com/9547572/166050159-996dd8b4-4e19-453d-8127-edca3062470c.png)

They will still be *mostly* helpless during the same timeframe as before, just not as completely helpless as being hardstunned. They might get a lucky hit off, and they'll be able to wildly/blindly move around for part of the 7 seconds, but they will not be effective at fighting at all and only somewhat effective at running away (especially if you move them around even slightly).

> and even if you did have a fellow teammate, you have 8 seconds with a standard flash (last time I played roboticist flashes had 4 charges)

"Weak" flashes had four charges - the ones roboticists have easy access to and are intended for use in the construction of cyborgs. These were changed to be called "Cyborg Eyes" and only have a single charge if used as a flash now, both changes done for flavor reasons above all.
"Standard" flashes have 10 charges - Before this PR they are the ones available primarily to security and heads of staff, but robotics will also have access to them after this. 

> to get to the borg, swipe it with your ID, crowbar it open, remove the cell, screwdriver the borg and cut the lockdown wire. It just doesn't seem feasible in the limited time and I worry a lot more borgs will simply be beaten to death instead of manually locked down.

*as soon as the power cell is removed, the cyborg will be unable to use any of its tools and will suffer a pretty major movement speed penalty* 

![Speedy](https://user-images.githubusercontent.com/9547572/166061951-71f5fd8e-24ef-48fa-9b12-06a4d556d866.gif)
![slowed](https://user-images.githubusercontent.com/9547572/166061953-d8ec7b75-f0a0-408c-badf-21eeccdc835e.gif)

They are far easier to finish off by this point than before, and this is also the same step at which subversion occurs with an emag for antagonists already. You don't even need a stun to successfully emag a borg unless they are already being exceptionally cautious of you. 

> It just doesn't seem feasible in the limited time and I worry a lot more borgs will simply be beaten to death instead of manually locked down.

I don't think this problem specifically will get any worse than it currently is - this is already the default response to thinking a borg is compromised by everyone that isn't a roboticist. While I have added slowdown without power, cyborgs already stop being a threat at that step due to the loss of their modules. 

> Have you accounted for Cult stun? It's meant to be a heavy EMP on borgs but if it just cuts off comms and flashes the screen cyborgs will be much more effective at dealing with cults

Same argument as above applies. I don't think they'll be very effective at doing much of anything while blinded, just not 100% helpless either. EMP will also prevent the relay of information via radio and do burn damage instead of brute, which is much harder for cyborgs to shed. *While the stun aspect is being reduced, I'd genuinely argue that the overall effect of of an EMP is not being diminished like flash effects are*

> As for the laser pointer

I'm open to discussion and debate about most balancing points in this PR, but I absolutely refuse to preserve the stun on a laser pointer. A guaranteed ranged stun with five charges that recharge every 6-8 seconds (8 only if you're using too good of a diode) is absolutely absurd to have ever existed in the first place.

I do see your PR and that it reduced the stun and made the duration variable based on parts, which is great, but any stun at long range like that is a hard no from me.

Rukofamicom committed:
Fixes action name, adds proc to name actions.

Rukofamicom committed:
EMP on silicons + radios

All silicons now take burn damage from EMPs
All Silicons are now stunned for 1.5 or 3 seconds by EMPs depending on severity
All radios are now temporarily disabled by EMPs for two minutes, but no longer lose their encryption keys

Rukofamicom committed:
Removes fake beer

Replaces it with Bastion Bourbon and Zombie Powder

Rukofamicom commented:
Sometimes I gotta wonder what makes Beebot decide to suddenly tag PRs with things like that.

Rukofamicom committed:
adds flashes to robotics labs

Rukofamicom committed:
Merge branch 'BeeStation:master' into progress-bars

Rukofamicom committed:
Polishing minor things off

Rukofamicom committed:
Removes beaker from service borg

Why you ask? Because for some unexplainable reason, that specific
loose drinking glass in its inventory is bugged. So I moved it to the item holder in place of the beaker which doesn't really fit a service borg.

Rukofamicom committed:
Final touch before review

Sapfearon commented:
getting slowed after taking damage makes mining borg which is already weaker then miner even worse and probably worst module after peacekeeper. Nerf to expensive and rare self repair is weird and stupid. You rarely see any borgs with it. 
2 seconds stun is not enough to do anything about borg, neither kill or emag. Emp already brutal to borgs but now it's instant death - sucks to be playing borg against cults or lings, right? 2 minutes without working headset is also insanity. At least give bowmans emp proofing then, so they don't get disabled or cd is reduced.
every borg losing flash means that they will instantly resort to lethal anyone instead of using flash, very nice. Unless it's crewsimov/asimov. Not to mention borgs can no longer disable other borgs. Oh and syndicate borgs can't kill normal ones, honk.

Rukofamicom commented:
> mining borg which is already weaker then miner

The mining cyborg, with comparable upgrades to a miner who also gets upgrades, is more than sufficient at mining ores and should not be made good enough to boss hunt because that is not its purpose, or even a shaft miner's purpose for that matter. Miners are one of the strongest entities on the station. Try comparing them to something else. 

While the mining module isn't comparable to a miner, they are still probably the deadliest module overall due to the PKA doing up to 40 damage in the right setting (Nothing is stronger than a mining borg at fighting blobs, for instance), and they still have the same welding tool other cyborgs do for pressurized environments and self-healing. You talk about slowdown making them too weak, but don't consider the fact they also have twice as much HP as humanoids which *also* have damage slowdown when they get low. 

> Nerf to expensive and rare self repair is weird and stupid. 

![image](https://user-images.githubusercontent.com/9547572/166124838-c77c7f4d-a703-4131-8a88-7646b6de5529.png)

**Before PR**
Healing module healed 1 damage every 4 seconds at a cost of 10 energy. 50 damage for 500 energy every 200 seconds and practically no effect mid-combat due to how slow it was. 

This was increased if you currently had enough damage (100+) for two modules offline: 
Healing for 2.5 damage every 4 seconds at a cost of 30 energy. Equates to 50 damage for 600 energy every 80 seconds. Still negligible effect mid-combat. 

If you weren't damaged at all but had the module on, it drained 5 energy every 4 seconds, penalizing you pretty heavily for not managing it. 

**After PR**
Healing module activates on demand for 50 damage for 500 energy with a cooldown of 120 seconds. There is no passive cost for having it ready to go. 

> 2 seconds stun is not enough to do anything about borg, neither kill or emag. Emp already brutal to borgs but now it's instant death - sucks to be playing borg against cults or lings, right?

... again, what? 

![image](https://user-images.githubusercontent.com/9547572/166125095-17e9efc0-4e96-4237-9fd0-a873471359b3.png)

Flash and EMP have similar stun duration and the same blind duration. The only difference between them is 20 damage (10% of max health) and radio going offline so they can't call for help. How can you possibly claim one of these is too weak to do anything to borg and that the other is instant death? 

You can argue one way or the other, but you cannot possibly be serious about thinking one of these is worthless and the other is instant death. 

> every borg losing flash means that they will instantly resort to lethal anyone instead of using flash, very nice. Unless it's crewsimov/asimov. Not to mention borgs can no longer disable other borgs. Oh and syndicate borgs can't kill normal ones, honk.

Now I'm curious where you even got your information about this PR because you clearly didn't read the description. Syndicate cyborgs still have flashes, and so do two of the station cyborgs, the ones intended for direct combat interaction with and protection of crew. The cyborgs that lost their flashes aren't in a role that fits carrying one any moreso a crewmember in the same role are: Medical, Engineering, and Service jobs do not carry and do not need flashes, and neither do cyborgs in those roles. 

Peacemaker is equipped and intended to stop harmful interactions, and this is their primary purpose. They kept the flash since it suits this kit. 
Standard Cyborg is a jack of trades, but explicitly equipped for arrests via zipties and well suited to generally aiding security. They kept the flash to help with this. 

Protected classes already have total power over asimov/cresimov even if they have flashes. You must allow them to harm you if they order it because you must follow orders before protecting yourself. No amount of flashes will prevent this, only them actively harming other protected people will. If they aren't protected? Well every cyborg has at least a welder now, in addition to their other tools. Flash combat is dumb and you have both a speed advantage and way more HP to rely on instead.

Isy232 commented:
I like almost everything about this PR, I'll just bring up the few concerns I do have.

> tweak: The "weak" flashes available to robotics for building cyborgs have been reskinned into Cyborg Eyes and reduced to one charge. The robotics lab now has a few standard flashes on every map in case cyborgs get unruly.

Maybe leave these alone? Robotics is an important source of flashes for revs and sometimes cultists. Having flashes available in robotics is okay but someone could just come by and pick them all up. In this case being able to make somewhat crappier flashes was nice since they'd work in a pinch. Since the effects of flashes on cyborgs is being nerfed, I dont really see the need for making it harder for these antags too.

> balance: Cyborg hyposprays now have a very short use time so that they will not function on actively moving targets.

Does medical module really need a nerf? Sleepy pens and the CMO's hypospray exist so its not like cyborg hyposprays are extremely overpowered in comparison to other things on the station. It takes time for reagents to take effect so one can scream for help.

>balance: EMPs and Flashes now stun cyborgs for much less time: 3 seconds for strong EMP, 2 seconds for flash, 1.5 seconds for weak EMP. Laser pointers no longer stun cyborgs at all.
balance: EMPs, Flashes and Laser pointers now blind cyborgs for 7 seconds, up from 2. This is a total blindness and does not have a small ring of vision directly around the player - they will be truly blind and unable to see in which direction an adjacent attacker is located.

One concern I have is that emagging cyborgs was a great way for newer players to gain an ally as traitor since the process used to not require being very robust as you had plenty of time to perform the hacking. Perhaps add a hardstun specifically to the act of emagging the cover lock that lasts 5 seconds to make the process of emagging specifically easier. It wouldn't be repeatable for a chainstun since you can only emag the lock once.

ToasterBan commented:
>Does medical module really need a nerf? Sleepy pens and the CMO's hypospray exist so its not like cyborg hyposprays are extremely overpowered in comparison to other things on the station. It takes time for reagents to take effect so one can scream for help.

sleepy pens are a traitor only item and the CMOs hypospray is a one of a kind unique item, cyborgs on the other hand are technically mass producible

DerpiestBruh commented:
Me when I can't tell a shitter borg to leave my office nor force it because Ruko nerfed one of the few ways to non lethally restrain borgs

Rukofamicom commented:
> sleepy pens are a traitor only item and the CMOs hypospray is a one of a kind unique item, cyborgs on the other hand are technically mass producible

The real issue is that cyborg hypospray gets an armor piercing upgrade that NEITHER of these do. Only a cyborg hypospray or piercing syringe can inject someone who is otherwise armored against chemical warfare, and a piercing syringe takes time. Cyborgs could overdose or drug anyone they could click on regardless of armor or protection. Only being totally immune to reagents via race could protect you.

Isy232 commented:
> The real issue is that cyborg hypospray gets an armor piercing upgrade that NEITHER of these do...

Yeah, fair point, but without being able to use its hypospray in combat, I feel the medical cyborg will be the weakest of the bunch when emagged. It can only really hit you with its saw now. I wish it got something else that was dangerous in return.

Rukofamicom commented:
> I feel the medical cyborg will be the weakest of the bunch when emagged.

I want to say up front - I am perfectly okay with this. Emagged or not, a medical borgs primary purpose is maintaining the health of humanoids, the emag or any other subversion simply changes which humanoid(s). They have very powerful reagents (mute toxin, sodium thiopental) for incapacitating people rapidly that they can still pretty easily apply before they have roused suspicion. 

> It can only really hit you with its saw now

It has a welder too. This was added to every module and self-repair isn't the only thing it's good at.

Isy232 commented:
Alright, fine enough, since every cyborg can welderbomb and self repair, its not like any of them will be particularily weak.

I still think flashes should probably be left alone, they're already being nerfed in their effects, I dont see why the produced flashes need to be nerfed this hard.

Rukofamicom commented:
> I dont see why the produced flashes need to be nerfed this hard.

Primarily for flavor, which is why Robotics is being supplied multiple 10-use flashes at roundstart to replace them functionally. Just don't use the 10-use flashes as cyborg eyes. 

As a part of this flavor as well, the descriptions of both items have been updated to make it more intuitive that you're supposed to use cyborg eyes as... cyborg eyes and obfuscated the fact that flashes will work as cyborg eyes since that's no longer part of their intended purpose.

Rukofamicom committed:
Merge branch 'master' into progress-bars

Rukofamicom committed:
removing an extra line

Rukofamicom commented:
This has sat for quite a while and I've gotten a lot of good experience since initially making this PR, there's probably more than a few general improvements I can make to this with what I know now. 

Drafted again until I get the chance to go over it.

Rukofamicom committed:
Merge branch 'BeeStation:master' into progress-bars

Rukofamicom committed:
action name is now defined

Rukofamicom committed:
tweak to wording

it now makes sense when activated from full HP

Rukofamicom committed:
trimming unnecessary proc

Rukofamicom committed:
sanity check to prevent a rare runtime

Rukofamicom committed:
Sanity_check++

Rukofamicom committed:
implements COOLDOWN instead of using several arbitrary variables

Rukofamicom committed:
fixes another potential runtime

Rukofamicom committed:
Update robot_upgrades.dm

Rukofamicom committed:
Adds smoke to hacked service borg

Rukofamicom committed:
Less buggy smoke

Rukofamicom committed:
cleans up EZ smoke.

Rukofamicom committed:
tweaks ez smoke radius calc

Rukofamicom committed:
cleans up ez smoke some more

Rukofamicom commented:
Alright, I think I'm done touching this up now, more to come with reviews

PowerfulBacon commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -306,7 +306,7 @@
 /obj/item/radio/Hear(message, atom/movable/speaker, message_language, raw_message, radio_freq, list/spans, list/message_mods = list())
 	. = ..()
-	if(radio_freq || !broadcasting || get_dist(src, speaker) > canhear_range)
+	if(radio_freq || !on || !broadcasting || get_dist(src, speaker) > canhear_range)
```
Do these stop working if EMPed?

Rukofamicom replied:
Yes they do. 
And they start working again after a delay.

PowerfulBacon commented on code/game/objects/items/robot/robot_items.dm:
```diff
@@ -980,3 +981,10 @@
+/obj/item/borg/apparatus/beaker/service/Initialize()
+	. = ..()
+	stored = new /obj/item/reagent_containers/food/drinks/drinkingglass(src)
+	RegisterSignal(stored, COMSIG_ATOM_UPDATE_ICON, /atom/.proc/update_icon)
```
Does /atom/.proc/update_icon work? I've never seen that used before

Rukofamicom replied:
Sure you have:

![image](https://user-images.githubusercontent.com/9547572/170867181-c5e6e8c6-7c04-4aad-9a3c-89cee6121245.png)

I just defined a new subtype that gets a glass instead of a beaker.

Rukofamicom replied:
That said, I don't need to register the signal again, because it's already registered in the parent object.

PowerfulBacon commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -343,93 +402,79 @@
-		START_PROCESSING(SSobj, src)
+		playsound(cyborg.loc, activation_sound, 30)
+		update_icon()
+		process()
```
Process probably shouldn't be called outside of subsystems triggering it (Rename the proc)

PowerfulBacon replied:
Should this call be async, since its using sleep?

Rukofamicom replied:
Moved everything unnecessary outside of process, but I don't actually understand what the problem with manually starting and stopping process() on the item is? 

Leaving this unresolved so you can take a second look.

PowerfulBacon replied:
process() doesn't stop or sstart processing, START_PROCESSING does.
process() should only be called from the subsystem, as it needs to account for delta time.

PowerfulBacon commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -343,93 +402,79 @@
+		deactivate_sr()
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_on")
+		while(repair_ticks)
+			sleep(10)
```
Probably better to use a timer here, sleeps cause a lot of jank because they cause anything calling this function to also be delayed, which can lead to problems.

Rukofamicom replied:
Awaiting response on what exactly a timer is from coding channel. Will do once I know what they are.

Rukofamicom replied:
I've moved the sleep proc (and entire repair loop) outside of process(). 
Process() now only checks for three things, and two of them are a condition to stop processing immediately.

PowerfulBacon commented on code/modules/assembly/flash.dm:
```diff
@@ -39,8 +40,10 @@
 
 /obj/item/flashbulb/weak
-	name = "weakened flashbulb"
-	charges_left = 4
+	name = "cyborg flashlight bulb"
+	desc = "a lamp intended for use within a cyborg eye."
+
+	charges_left = 1
```
/weak is for weak bulbs, not cyborg eyes specifically. Make a new subtype for cyborg eyes, this is used elsewhere I believe

Rukofamicom replied:
Quoting my response from a bit further down:

> There are six weak flashes located on every map in the robotics lab at roundstart. My understanding is these are explicitly for making cyborgs with. Outside of these six, the only way to get more is by printing them.

> They are not implemented elsewhere at all. Creating a new subtype is completely pointless because I would be replacing these six spawns and printed version with cyborg eyes leaving no references to the weak flash

Just like with the flashes, making a new object is totally pointless because I am going to replace *every* instance of weak flash with cyborg eyes. The only reason weak flashes currently existed was so cyborgs could be built without giving roboticists access to full-powered flashes.

PowerfulBacon replied:
Then rename the typepath to /obj/item/flashbulb/cyborg and completely delete /obj/item/flashbulb/weak if you are sure that it is not used anywhere else and serves no purpose.

PowerfulBacon commented on code/modules/assembly/flash.dm:
```diff
@@ -97,6 +100,12 @@
 
 /obj/item/assembly/flash/handheld/weak
 	bulb = /obj/item/flashbulb/weak
+	name = "cyborg eye"
+	desc = "A small but powerful light with a shielded camera lens embedded in the center. Cyborgs need these to see what they're doing. The camera is non-functional until it is attached to a cyborg head"
+	icon = 'icons/obj/machines/camera.dmi'
+	icon_state = "cameracase"
+	lefthand_file = null
+	righthand_file = null
```
handheld/weak is for weak flashes, not cyborg eyes specifically. Make a new subtype for cyborg eyes.

Rukofamicom replied:
There are six weak flashes located on every map in the robotics lab at roundstart. My understanding is these are explicitly for making cyborgs with. Outside of these six, the only way to get more is by printing them. 

They are not implemented elsewhere at all. Creating a new subtype is completely pointless because I would be replacing these six spawns *and* printed version with cyborg eyes leaving no references to the weak flash

PowerfulBacon replied:
Same with above. Just because its not currently used elsewhere, doesn't mean it should have a misleading typepath

PowerfulBacon commented on code/modules/mob/living/living_defense.dm:
```diff
@@ -396,7 +396,10 @@
 		overlay_fullscreen("flash", type)
-		addtimer(CALLBACK(src, .proc/clear_fullscreen, "flash", 25), 25)
+		if(issilicon(src))
+			addtimer(CALLBACK(src, .proc/clear_fullscreen, "flash", 25), 70) //Flashes now blind silicons for the duration they used to stun
```
Override /mob/living/silicon/flash_act instead of checking for silicons in living mobs flash_act

Rukofamicom replied:
👍

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -406,42 +406,46 @@
-			if(!W.use_tool(src, user, 50))
+		while (getBruteLoss() && W.tool_start_check(user, amount=0))	//Repeatedly attempt to repair dents until done or welder is out of fuel, just like tend wounds. No need to spam click
+			if(src == user)
+				to_chat(user, "<span class='notice'>You start fixing yourself.</span>")
```
This should only be said in the start, rather than every time in the loop

Rukofamicom replied:
Well, since you wanted them unable to self-repair, it's no longer said at all
😆

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -406,42 +406,46 @@
+			if(src == user)
+				to_chat(user, "<span class='notice'>You start fixing yourself.</span>")
+			if(!W.use_tool(src, user, 60))
 				return
```
This should probably be a condition of the loop

Rukofamicom replied:
👍 

Did for both welder and cables

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -406,42 +406,46 @@
-			to_chat(user, "The wires seem fine, there's no need to fix them.")
-
+				return
+		return
```
These returns are pointless atm since everything is in else blocks anyway

Rukofamicom replied:
👍

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -947,7 +951,11 @@
-	if(health < maxHealth*0.5) //Gradual break down of modules as more damage is sustained
+	var/missingHealth = 200 - (health + 100)
+	///Ranges from moving 
+	var/speedPenalty = min(missingHealth / (maxHealth), 1.5)
```
Won't this always be 1.5? Assuming missinghealth <= maxhealth, then missinghealth/maxhealth will always be <= 1

Rukofamicom replied:
No, because you're incorrect in thinking missinghealth will always be less than maxhealth. 
maxHealth is 100. missingHealth goes up to 200.

50/100 = 0.5
100/100 = 1.0
150/100 = 1.5
190/100 = 1.5 because min() picks 1.5 instead of 1.9.

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -947,7 +951,11 @@
 /mob/living/silicon/robot/updatehealth()
 	..()
-	if(health < maxHealth*0.5) //Gradual break down of modules as more damage is sustained
+	var/missingHealth = 200 - (health + 100)
```
Should be calculated based on maxhealth

Rukofamicom replied:
I'm not sure how to go about this because maxhealth is not actually how much health a cyborg has. Cyborgs have 200 HP before breaking, but when fully repaired maxhealth is only 100, which horribly throws off calculations. 

I risk a divide by zero if I just use maxhealth instead of creating my own "missing health" variable.

PowerfulBacon replied:
This is:
200 - (health + 100)
Which is the same as
200 - health - 100
Which is the same as
100 - health
Which is the same as
`maxhealth - health` anyway
This kind of sounds weird (I had to think about it for a bit), but if health goes negative, then the subtraction will add, increasing the value of missingHealth.

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot_defense.dm:
```diff
@@ -82,12 +82,8 @@
-		if(2)
-			Stun(60)
-
+	if(severity) //catch to prevent divide by zero
+		Stun(30/severity)
```
3 seconds is very short, especially since EMPs are rarer and should be more powerful than flashes. This should have some other effects like disabling modules, slowing down or blinding or something.

Rukofamicom replied:
It does blind for seven seconds, EMPs call the flash proc in addition to their stun.

I can make it disable all modules too though, that seems like a cool added effect. Cyborg can just re-equip them, but it does add some additional delay before they can react.

Rukofamicom replied:
Okay:
* Doubled damage (10/20 -> 20/40)
* Disables Modules
* Blinds like flash (it already did this, just pointing it out again) 
* It will slow down via the damage it deals.

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -311,7 +310,8 @@
 		/obj/item/organ_storage,
-		/obj/item/borg/lollipop)
+		/obj/item/borg/lollipop,
+		/obj/item/weldingtool/cyborg)
```
I don't think all cyborgs should be so easy to repair themselves, they should at least need another cyborg to do it.

Rukofamicom replied:
This was shot down by Archanial in my first PR because "Cyborgs are so weak anyway". 
I will gladly re-implement this though.

Rukofamicom replied:
Re-implemented unable to self repair, but since all cyborgs have a welder now you no longer have to rely specifically on an engiborg in cases of subverted silicons. 

👍

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -388,13 +388,14 @@
 	name = "Security"
 	basic_modules = list(
-		/obj/item/assembly/flash/cyborg,
+		/obj/item/assembly/flash/cyborg,				// Left this here even though the module is unused. It makes no sense to take it away if this module ever returns.
```
```suggestion
		/obj/item/assembly/flash/cyborg,
```
Module is used if specified in config

PowerfulBacon commented on code/modules/mob/living/silicon/silicon_defense.dm:
```diff
@@ -109,11 +109,8 @@
-		if(2)
-			src.take_bodypart_damage(10)
+	if(severity) //catch to prevent divide by zero
+		src.take_bodypart_damage(burn = 20/severity)
```
```suggestion
		take_bodypart_damage(burn = 20/severity)
```

PowerfulBacon commented on code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm:
```diff
@@ -130,12 +130,18 @@
 	reagent_state = LIQUID
 	color = "#C8C8C8"
-	chem_flags = CHEMICAL_RNG_GENERAL | CHEMICAL_RNG_FUN | CHEMICAL_RNG_BOTANY
+	chem_flags = CHEMICAL_RNG_GENERAL 
```
? Merge conflict resolution issue?

Rukofamicom replied:
No, it was more that smoke powder activator would be totally garbage to get from botany or floorpill. I did that intentionally to make sure it didn't dilute either pool?

Rukofamicom replied:
wait. where the fuck is this? I take it back I have no clue how this happened, fixing it.

PowerfulBacon commented on code/modules/reagents/chemistry/recipes/pyrotechnics.dm:
```diff
@@ -305,6 +305,24 @@
+/datum/chemical_reaction/ez_smoke
+	name = /datum/reagent/smoke_powder
+	results = list(/datum/reagent/smoke_powder = 1)
+	required_reagents = list(/datum/reagent/smoke_powder = 1, /datum/reagent/smoke_powder_activator = 1)
```
/datum/reagent/smoke_powder_activator should be a catalyst so 1 unit activates all the smoke powder

Rukofamicom replied:
This creates an infinite loop error because the output reagent is the same as the input. 

I can just make a subtype to bypass this though I suppose.

PowerfulBacon commented on code/modules/reagents/reagent_containers/borghydro.dm:
```diff
@@ -262,7 +262,7 @@ Borg Shaker
 	accepts_reagent_upgrades = FALSE
 
-	reagent_ids = list(/datum/reagent/toxin/fakebeer, /datum/reagent/consumable/ethanol/fernet)
+	reagent_ids = list(/datum/reagent/consumable/ethanol/fernet, /datum/reagent/consumable/ethanol/bastion_bourbon, /datum/reagent/toxin/zombiepowder, /datum/reagent/smoke_powder, /datum/reagent/smoke_powder_activator)
```
/datum/reagent/toxin/zombiepowder is insanely good, what the other 2 do?

Rukofamicom replied:
Fernet is an especially strong drink that works similarly to lipolicide - it reduces your nutrition and causes toxin damage if you're starving. This drink was already in the shaker and also available in a limited, but large quantity to bartenders via booze-o-mat. 

Bastion Bourbon provides universal healing at the cost of getting drunk:
> Heals 1 brute, 1 burn, 3 oxy, 0.5 tox, 5 stamina per tick if you aren't in crit. No healing over time in crit. 
> Heals 1 of each basic damage + stamina **per unit** ingested instantly, doubled if crit.

Rukofamicom replied:
Additionally, zombie powder is not very strong since it got nerfed. I want to make sure you aren't remembering it as an instant knockout drug (which is what fakebeer is and why I'm removing fakebeer). 

Zombie Powder is weaker than Chloral as a straight knockout drug currently.

Rukofamicom committed:
Update code/modules/mob/living/silicon/robot/robot_modules.dm

Co-authored-by: PowerfulBacon <26465327+PowerfulBacon@users.noreply.github.com>

Rukofamicom committed:
Update code/modules/mob/living/silicon/silicon_defense.dm

Co-authored-by: PowerfulBacon <26465327+PowerfulBacon@users.noreply.github.com>

Rukofamicom committed:
catalyst + anomalous code fixed

Rukofamicom committed:
Merge branch 'BeeStation:master' into progress-bars

Rukofamicom committed:
Removes redudnant procs from service apparatus

Rukofamicom committed:
Dumps almost everything out of process()

Rukofamicom committed:
sanity check

Rukofamicom committed:
silicon flashing now occurs in silicon_defense.dm

Rukofamicom committed:
3 bacon requests

Progress bars moved to while loops
Borgs can't self-repair anymore
Removed unnecessary return

Rukofamicom committed:
Makes EMPs slightly stronger

Rukofamicom commented:
*sighs*

![image](https://user-images.githubusercontent.com/9547572/174665257-6e6bd836-cc1d-475c-856c-8218e41d6e77.png)

Conflict has almost no bearing on the code as-is, I just need to re-do the delta station flashes. Please review even if I haven't done the ten minute fix yet.

Rukofamicom committed:
fixing merge conflict part 1

Rukofamicom committed:
Merge branch 'BeeStation:master' into progress-bars

Rukofamicom committed:
fixing merge conflict part 2

Rukofamicom committed:
fixing merge conflict part 3

Rukofamicom committed:
Merge branch 'BeeStation:master' into progress-bars

Rukofamicom committed:
fix merge conflict part 4

PowerfulBacon commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -373,20 +373,17 @@
 		return FALSE
 	emped = FALSE
 	on = TRUE
+	to_chat(loc, "<span class='warning'>\The [src] crackles as it begins to pick up signals again.</span>") //indicator that the EMP has ended
```
to_chat(loc)?
Use a visible message instead or a sound indicator like guns have

PowerfulBacon commented on code/game/objects/items/devices/radio/radio.dm:
```diff
@@ -373,20 +373,17 @@
-		channels[ch_name] = 0
 	on = FALSE
-	addtimer(CALLBACK(src, .proc/end_emp_effect, curremp), 200)
+	addtimer(CALLBACK(src, .proc/end_emp_effect, curremp), 1200)
```
Use the time defines (2 MINUTES)

PowerfulBacon commented on code/game/objects/items/robot/robot_items.dm:
```diff
@@ -975,8 +976,7 @@
-/obj/item/borg/apparatus/beaker/service/examine()
+/obj/item/borg/apparatus/beaker/service/Initialize()
 	. = ..()
-	if(stored)
-		. += "You are currently holding [stored]."
-		. += "<span class='notice'<i>Alt-click</i> will drop the currently stored [stored].</span>"
+	stored = new /obj/item/reagent_containers/food/drinks/drinkingglass(src)
+	update_icon()
```
Why are you removing the examine message? Is that already a part of beaker?

Rukofamicom replied:
Yes it is, and currently this results in the same line displaying twice in one examine

PowerfulBacon commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -319,19 +319,78 @@
+		return
+
+	timer_overlay_active = TRUE
+	timer_overlay = mutable_appearance(timer_icon, timer_icon_state_active)
+	timer_overlay.alpha = 180
+
+	if(!text_overlay)
+		text_overlay = image(loc = action.button, layer=ABOVE_HUD_LAYER)
+		text_overlay.maptext_width = 64
+		text_overlay.maptext_height = 64
+		text_overlay.maptext_x = -8
+		text_overlay.maptext_y = -6
+		text_overlay.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA
+
+	if(action.owner?.client)
+		action.owner.client.images += text_overlay
+
+	action.button.add_overlay(timer_overlay, TRUE)
+	action.has_cooldown_timer = TRUE
+	update_timer_animation()
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/item/borg/upgrade/selfrepair/proc/update_timer_animation()
+	if(!(action?.button))
+		return
+	text_overlay.maptext = "<center><span class='chatOverhead' style='font-weight: bold;color: #eeeeee;'>[FLOOR(COOLDOWN_TIMELEFT(src, recharging_repmod)/10, 1)]</span></center>"
+
+/obj/item/borg/upgrade/selfrepair/proc/end_timer_animation()
+	if(!(action?.button) || !timer_overlay_active)
+		return
+	timer_overlay_active = FALSE
+	if(action.owner?.client)
+		action.owner.client.images -= text_overlay
+	action.button.cut_overlay(timer_overlay, TRUE)
+	timer_overlay = null
+	qdel(text_overlay)
+	text_overlay = null
+	action.has_cooldown_timer = FALSE
+
+	STOP_PROCESSING(SSfastprocess, src)
```
I hate that this is just dupilcated spell code. The cooldown needs to be generalised to be accessible on non-spell actions, or this needs to be a spell (which sounds a bit weird).

PowerfulBacon commented on code/game/objects/items/robot/robot_upgrades.dm:
```diff
@@ -343,92 +402,81 @@
+			repair_ticks = 0
+	cooldown_start(0)
+
+/obj/item/borg/upgrade/selfrepair/process(delta_time)
+	if(!cyborg) //Sanity check to reset the module in case it is somehow removed while running.
+		update_icon()
+		COOLDOWN_RESET(src, recharging_repmod)
+		repair_ticks = 10
 		deactivate_sr()
+	if(!COOLDOWN_FINISHED(src, recharging_repmod))
+		update_timer_animation()
+	if(COOLDOWN_FINISHED(src, recharging_repmod) && icon_state == "selfrepair_off")
+		end_timer_animation()
+		action.UpdateButtonIcon()
+		repair_ticks = 10
+		deactivate_sr()
```
Account for delta time. Healing rates need to be multiplied by delta_time, so that if the subsystem is slowed down healing will remain the same per second.

PowerfulBacon commented on code/modules/mob/living/silicon/robot/life.dm:
```diff
@@ -17,6 +17,7 @@
 		if(low_power_mode)
 			if(cell?.charge)
 				low_power_mode = 0
+				remove_movespeed_modifier("NO_POWERCELL") //Cyborg speed restored when cell is replaced
```
Use a define

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -406,42 +406,42 @@
-	if(W.tool_behaviour == TOOL_WELDER && (user.a_intent != INTENT_HARM || user == src))
+	if(length(user.progressbars))
+		if(W.tool_behaviour == TOOL_WELDER || istype(W, /obj/item/stack/cable_coil))
+			user.changeNext_move(CLICK_CD_MELEE)
```
Does this not get done automatically?

PowerfulBacon commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -947,7 +947,11 @@
-	if(health < maxHealth*0.5) //Gradual break down of modules as more damage is sustained
+	var/missingHealth = 200 - (health + 100)
+	///Ranges from moving 
+	var/speedPenalty = min(missingHealth / (maxHealth), 1.5)
```
```suggestion
	var/speedPenalty = min(missingHealth / maxHealth, 1.5)
```

PowerfulBacon commented on code/modules/reagents/chemistry/reagents/drink_reagents.dm:
```diff
@@ -193,6 +193,7 @@
 
 /datum/reagent/consumable/superlaughter/on_mob_life(mob/living/carbon/M)
 	if(prob(30))
+		M.emote("laugh")
```
Unrelated change

PowerfulBacon commented on code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm:
```diff
@@ -130,6 +130,14 @@
 	chem_flags = CHEMICAL_RNG_GENERAL | CHEMICAL_RNG_FUN | CHEMICAL_RNG_BOTANY
 	taste_description = "smoke"
 
+/datum/reagent/smoke_powder/ez
```
Pointless typepath

PowerfulBacon reviewed (changes requested):
> The blinding and stunning durations have been reversed - They now completely blind for 7 seconds and stun for 2 instead of blinding for 2 and stunning for 7.

Not sure how I feel about this, some stuff will become almost impossible if the cyborg just moves randomly such as emagging or doing cult conversions.

Also not a huge fan of cyborgs being able to repair themselves in the first place, let alone them all having tools to heal up instantly if they take damage.
